typing/typecore.ml

```diff
@@ -3001,10 +3001,18 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 sargs =
                 if not optional && is_optional l' then
                   Location.prerr_warning sarg.pexp_loc
                     (Warnings.Nonoptional_label (Asttypes.string_of_label l));
-                remaining_sargs, Some (sarg, l')
+                remaining_sargs, Some (sarg, l'), visited, false
             | None ->
-                sargs, None
+                match TypeMap.find_opt ty_fun visited with
+                | Some sargs' when sargs = sargs' ->
+                    sargs, None, visited, true
```

Comment:
Rather than storing the list of `sargs`, I would propose to use a `TypeSet`, and make it empty whenever the list of syntactic arguments decreases.